src/analysis/normed_space/real_inner_product.lean

```diff
@@ -0,0 +1,174 @@
+  -/
+
+noncomputable theory
+
+universes u v w
+
+variables {α : Type u} {F : Type v} {G : Type w}
+
+class has_inner (α : Type*) := (inner : α → α → ℝ)
+
+export has_inner (inner)
+
+/--
+An inner product space is a real vector space with an additional operation called inner product.
+Inner product spaces over complex vector space will be defined in another file.
+-/
+class inner_product_space (α : Type*) [add_comm_group α] [vector_space ℝ α] extends has_inner α :=
+(comm      : ∀ x y, inner x y = inner y x)
+(nonneg    : ∀ x, 0 ≤ inner x x)
+(definite  : ∀ x, inner x x = 0 → x = 0)
+(add_left  : ∀ x y z, inner (x + y) z = inner x z + inner y z)
+(smul_left : ∀ x y r, inner (r • x) y = r * inner x y)
+
+variables [add_comm_group α] [vector_space ℝ α] [inner_product_space α]
+
+section basic_properties
+
+lemma inner_comm (x y : α) : inner x y = inner y x := inner_product_space.comm x y
+
+lemma inner_self_nonneg {x : α} : 0 ≤ inner x x := inner_product_space.nonneg _
+
+lemma inner_add_left {x y z : α} : inner (x + y) z = inner x z + inner y z :=
+inner_product_space.add_left _ _ _
+
+lemma inner_add_right {x y z : α} : inner x (y + z) = inner x y + inner x z :=
+by { rw [inner_comm, inner_add_left], simp [inner_comm] }
+
+lemma inner_smul_left {x y : α} {r : ℝ} : inner (r • x) y = r * inner x y :=
+inner_product_space.smul_left _ _ _
+
+lemma inner_smul_right {x y : α} {r : ℝ} : inner x (r • y) = r * inner x y :=
+by { rw [inner_comm, inner_smul_left, inner_comm] }
+
+@[simp] lemma inner_zero_left {x : α} : inner 0 x = 0 :=
+by { rw [← zero_smul ℝ (0:α), inner_smul_left, zero_mul] }
+
+@[simp] lemma inner_zero_right {x : α} : inner x 0 = 0 :=
+by { rw [inner_comm, inner_zero_left] }
+
+lemma inner_self_eq_zero (x : α) : inner x x = 0 ↔ x = 0 :=
+iff.intro (inner_product_space.definite _) (by { rintro rfl, exact inner_zero_left })
+
+@[simp] lemma inner_neg_left {x y : α} : inner (-x) y = -inner x y :=
+by { rw [← neg_one_smul ℝ x, inner_smul_left], simp }
+
+@[simp] lemma inner_neg_right {x y : α} : inner x (-y) = -inner x y :=
+by { rw [inner_comm, inner_neg_left, inner_comm] }
+
+@[simp] lemma inner_neg_neg {x y : α} : inner (-x) (-y) = inner x y := by simp
+
+lemma inner_sub_left {x y z : α} : inner (x - y) z = inner x z - inner y z :=
+by { simp [sub_eq_add_neg, inner_add_left] }
+
+lemma inner_sub_right {x y z : α} : inner x (y - z) = inner x y - inner x z :=
+by { simp [sub_eq_add_neg, inner_add_right] }
+
+/-- Expand `inner (x + y) (x + y)` -/
+lemma inner_add_add_self {x y : α} : inner (x + y) (x + y) = inner x x + 2 * inner x y + inner y y :=
+by { simpa [inner_add_left, inner_add_right, two_mul] using inner_comm _ _ }
+
+/-- Expand `inner (x - y) (x - y)` -/
+lemma inner_sub_sub_self {x y : α} : inner (x - y) (x - y) = inner x x - 2 * inner x y + inner y y :=
+by { simp only [inner_sub_left, inner_sub_right, two_mul], simpa using inner_comm _ _ }
+
+/-- Parallelogram law -/
+lemma parallelogram_law {x y : α} :
```

Comment:
I feel the descriptive name `inner_add_add_self_add_inner_sub_sub_self` is terrible. But not sure about this